File: src/main/java/PrintInOrder.java
import java.util.concurrent.Semaphore;

public class PrintInOrder {

    static class SemaphoresFoo {

        private final Semaphore first = new Semaphore(1);
        private final Semaphore second = new Semaphore(1);
        private final Semaphore third = new Semaphore(1);

        public SemaphoresFoo() {
            try {
                first.acquire();
                second.acquire();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }

        public void first(Runnable printFirst) throws InterruptedException {
            third.acquire();
            printFirst.run();
            first.release();
        }

        public void second(Runnable printSecond) throws InterruptedException {
            first.acquire();
            printSecond.run();
            second.release();
        }

        public void third(Runnable printThird) throws InterruptedException {
            second.acquire();
            printThird.run();
            third.release();
        }
    }

    static class SynchronizedFoo {

        private boolean isFirst = false;
        private boolean isSecond = false;
        private boolean isThird = true;

        public SynchronizedFoo() {

        }

        public synchronized void first(Runnable printFirst) throws InterruptedException {
            while (!isThird) {
                wait();
            }
            printFirst.run();
            isFirst = true;
            notifyAll();
        }

        public synchronized void second(Runnable printSecond) throws InterruptedException {
            while (!isFirst) {
                wait();
            }
            printSecond.run();
            isSecond = true;
            notifyAll();
        }

        public synchronized void third(Runnable printThird) throws InterruptedException {
            while (!isSecond) {
                wait();
            }
            printThird.run();
            isThird = true;
            notifyAll();
        }
    }

}

File: src/main/kotlin/PrintInOrder.kt
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex

object PrintInOrderKt {
    class MutexFoo : OrderedPrinter {
        private val first = Mutex()
        private val second = Mutex(locked = true)
        private val third = Mutex(locked = true)

        override fun first(printFirst: Runnable) = runBlocking {
            first.lock()
            printFirst.run()
            second.unlock()
        }

        override fun second(printSecond: Runnable) = runBlocking {
            second.lock()
            printSecond.run()
            third.unlock()
        }

        override fun third(printThird: Runnable) = runBlocking {
            third.lock()
            printThird.run()
            first.unlock()
        }
    }

    class FlowFoo : OrderedPrinter {
        private val order = MutableStateFlow(Order.FIRST)

        override fun first(printFirst: Runnable) = runBlocking {
            order.first { it == Order.FIRST }
            printFirst.run()
            order.update { Order.SECOND }
        }

        override fun second(printSecond: Runnable) = runBlocking {
            order.first { it == Order.SECOND }
            printSecond.run()
            order.update { Order.THIRD }
        }

        override fun third(printThird: Runnable) = runBlocking {
            order.first { it == Order.THIRD }
            printThird.run()
            order.update { Order.FIRST }
        }

        private enum class Order {
            FIRST, SECOND, THIRD
        }
    }
}
